lib/src/config_document_parser.cc

```diff
@@ -0,0 +1,632 @@
+        path* last_path = nullptr;
+        bool last_inside_equals = false;
+        shared_node_list object_nodes;
+        shared_node_list key_value_nodes;
+        unordered_map<string, bool> keys;
+
+        if (had_open_curly) {
+            object_nodes.push_back(make_shared<config_node_single_token>(tokens::open_curly_token()));
+        }
+
+        while (true) {
+            shared_token t = next_token_collecting_whitespace(object_nodes);
+            if (t->get_token_type() == token_type::CLOSE_CURLY) {
+                if (_flavor == config_syntax::JSON && after_comma) {
+                    throw parse_exception(add_quote_suggestion(t->to_string(),
+                        "expecting a field name after a comma, got a close brace '}' instead", _equals_count > 0));
+                } else if (!had_open_curly) {
+                    throw parse_exception(add_quote_suggestion(t->to_string(),
+                           "unbalanced close brace '}' with no open brace", _equals_count > 0));
+                }
+                object_nodes.push_back(make_shared<config_node_single_token>(tokens::close_curly_token()));
+                break;
+            } else if (t->get_token_type() == token_type::END && !had_open_curly) {
+                put_back(t);
+                break;
+            } else if (_flavor != config_syntax::JSON && is_include_keyword(t)) {
+                shared_node_list include_nodes;
+                include_nodes.push_back(make_shared<config_node_single_token>(t));
+                shared_ptr<config_node_include> inc = parse_include(include_nodes);
+                object_nodes.push_back(inc);
+                after_comma = false;
+            } else {
+                key_value_nodes.clear();
+                shared_token key_token = t;
+                auto key_path = parse_key(key_token);
+                key_value_nodes.push_back(key_path);
+                shared_token after_key = next_token_collecting_whitespace(key_value_nodes);
+                bool inside_equals = false;
+
+                shared_node_value next_value = nullptr;
+                if (_flavor == config_syntax::CONF && after_key->get_token_type() == token_type::OPEN_CURLY) {
+                    // can omit the ':' or '=' befor an object value
+                    next_value = parse_value(after_key);
+                } else {
+                    if (!is_key_value_separator(after_key)) {
+                        throw parse_exception(add_quote_suggestion(after_key->to_string(),
+                        "Key '" + key_path->render() + "' may not be followed by token: " + after_key->to_string(),
+                        _equals_count > 0));
+                    }
+
+                    key_value_nodes.push_back(make_shared<config_node_single_token>(after_key));
+                    if (after_key->get_token_type() == token_type::EQUALS) {
+                        inside_equals = true;
+                        _equals_count++;
```

Comment:
I'm not sure if the compiler optimizes this when the return value is ignored, but semantically doing `++__equals_count` is less work because it increments then returns that value. `_equals_count++` potentially has to save a temporary, increment the original, and return the temporary. It's a good habit to prefer the prefix notation.